feat(website): getting started revamp (#1285)
cdk8s-team/cdk8s: docs/get-started/python.md

# Getting started with CDK8s for Python
In this guide, we'll walk you through the most essential concepts you need to know in order to build a Kubernetes application using CDK8s for Python.

## Prerequisites
1. CDK8s for Python supports Python versions 3.7+. Check your Python version:
```bash
python3 --version
```

## Setting up an environment
### Initial setup
To get setup, follow these general steps:

1. To begin, you’ll need a development environment with pipenv 2018.11.26+ installed. If you’re not familiar with this setup, use these instructions to [setup pipenv](https://pipenv.pypa.io/en/latest/install/).

> 2. Optionally, install an IDE extension to enable code autocompletion. The Code Whisperer [VS Code](https://marketplace.visualstudio.com/items?itemName=AmazonWebServices.aws-toolkit-vscode) and [JetBrains](https://plugins.jetbrains.com/plugin/11349-aws-toolkit) Extensions available in the AWS Toolkit support auto-completion for CDK8s and CDK8s+.

Reviewer: This makes it sound like code completion will only be available if they install an extension, and even more specifically an AWS extension. This isn't true. Code completion is available out of the box in all IDEs that support your language of choice. I would skip this entire thing.

Author: I'm okay to remove.

### Install the CLI
To initialize a CDK8s project and auto-generate Kubernetes manifests based on our code, we need the CDK8s CLI:

1. Run the following command to install the CLI using [Homebrew](https://brew.sh/). For more installation methods, see [Install the CDK8s CLI](cli/installation.md).
```console

> brew install cdk8s

Reviewer: This is the specific for mac, lets just use `npm install -g cdk8s-cli`

Author: Absent OS-type metrics, if more Typescript developers use CDK8s, then I think that makes sense to use npm.

   * The `synth` method is called on the app instance, which produces the required Kubernetes YAML manifest files based on the defined resources. Note that in this example, we haven't defined any resources within the `MyChart` constructor, so running the "cdk8s synth" command in the CLI would generate a blank Kubernetes manifest.

## Defining Kubernetes resources
Now, we are ready to define Kubernetes resources for our application. In this sample, we illustrate the creation of a basic Kubernetes Service for a sample application, beginning with the import of the k8s library from the imports package, which enables us to work with Kubernetes resources.

### Copy the code sample
1. Copy and paste the following code sample into the existing `main.py` file of your project.
```python
#!/usr/bin/env python
from constructs import Construct
from cdk8s import App, Chart
from imports import k8s

class MyAwesomeApp(Chart):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        # Define a Kubernetes Service

>         k8s.KubeService(self, "my-service",

Reviewer: Should we use a `KubeDeployment` maybe? a `KubeService` doesn't make much since on its own.

Author: That crossed my mind, but we're guiding them to the complete Deployment and Service in the "Hello" example as a next step. Don't think we want to go beyond the basics at this point, we're already pushing it with the docs under the sample.

Author: Improved the sample/descriptions to show how to pass a custom label parameter, demonstrating reuse.

Author: Updated to a Deployment, closing.

* An instance of the `MyAwesomeApp` class is created with the identifier "python-service".

## Generating Kubernetes manifests
After you have defined the Kubernetes resources for your project, you are ready to generate the Kubernetes manifest that will define your Service resource. 

### Run the synth command
1. Open a terminal and navigate to your project directory.
2. Run the [synth](https://cdk8s.io/docs/latest/cli/synth/) command. This command generates a Kubernetes manifest file in the `dist` folder of your project directory. The manifest file is based on the parameters specified in your application.
```console
cdk8s synth
```

### View the manifest
1. Open the `dist/go-service.k8s.yaml` file. You should see a manifest similar to the following:
```yaml
apiVersion: v1
kind: Service
metadata:
 name: go-service-service-c8d5f888
spec:
 ports:
   - port: 80
     targetPort: 8080
 selector:
   app: my-app
 type: LoadBalancer
```

## Wrap up

> - To continue your journey, develop a complete Kubernetes [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) and [Service](https://kubernetes.io/docs/concepts/services-networking/service/) using the [Hello, Kubernetes!](https://github.com/cdk8s-team/cdk8s/tree/master/examples/python/hello) sample application.

Reviewer: The new [examples repo](https://github.com/cdk8s-team/cdk8s-examples) is now ready btw.

Author: Awesome, I'll update.

# Getting started with CDK8s for Python
In this guide, we'll walk you through the most essential concepts you need to know in order to build a Kubernetes application using CDK8s for Python.

## Prerequisites
1. CDK8s for Python supports Python versions 3.7+. Check your Python version:
```bash
python3 --version
```

## Setting up an environment
### Initial setup
To get setup, follow these general steps:

1. To begin, you’ll need a development environment with pipenv 2018.11.26+ installed. If you’re not familiar with this setup, use these instructions to [setup pipenv](https://pipenv.pypa.io/en/latest/install/).

### Install the CLI
To initialize a CDK8s project and auto-generate Kubernetes manifests based on our code, we need the CDK8s CLI:

1. Run the following command to install the CLI using [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm). For more installation methods, see [Install the CDK8s CLI](cli/installation.md).
```console
npm install -g cdk8s-cli
```

### Create a project

> Next, we’ll initialize a project to create the directory structure, install the CDK8s and CDK8s+ packages and all of the dependencies required to run an application using the [init command](https://cdk8s.io/docs/latest/cli/init/). 

Reviewer: ```suggestion
Next, we’ll initialize a project to create the directory structure and install the necessary dependencies using the [init command](https://cdk8s.io/docs/latest/cli/init/). 
```

Author: Updated language.

> 1. In a terminal window, run the `init` command in an empty directory: 

Reviewer: ```suggestion
1. In a terminal window, run the `cdk8s init` command in an empty directory: 
```

Author: Since we already linked to/indicated we're using the init command, changed to "run the following command"

```console
cdk8s init python-app
```

2. In your IDE, open the `main.py` file. 
```python
#!/usr/bin/env python
from constructs import Construct
from cdk8s import App, Chart

class MyChart(Chart):
   def __init__(self, scope: Construct, id: str):
       super().__init__(scope, id)

       # define resources here

app = App()
MyChart(app, "python")

app.synth()
```

> This sample shows the basic structure of a CDK8s application with the essential libraries: **constructs** and **cdk8s**. These libraries supply the foundational classes and methods required for working with CDK8s. It includes the following components:
> 
>    * A custom `MyChart` class inherits from the [Chart](https://cdk8s.io/docs/latest/reference/cdk8s/python/#chart) base class provided by the cdk8s library, serving as a representation of the Kubernetes resources to be generated and managed.
>    * The `__init__` method within the `MyChart` class is responsible for initializing the base class ([Chart](https://cdk8s.io/docs/latest/reference/cdk8s/python/#chart)) and specifying the Kubernetes resources. This method is invoked when creating an instance of the class.
>    * An instance of the [App](https://cdk8s.io/docs/latest/reference/cdk8s/python/#app) class signifies the primary entry point of the CDK8s application and oversees the application's lifecycle and resources.
>    * An instance of the `MyChart` class is created by passing the app instance and a string identifier, "python", as arguments. This action generates and registers the Kubernetes resources defined in the `MyChart` class within the application. Note that in this example, the "python" string identifier is autogenerated based on the current directory name.
>    * The `synth` method is called on the `app` instance, which produces the required Kubernetes YAML manifest files based on the defined resources. Note that in this example, we haven't defined any resources within the `MyChart` constructor, so running the "cdk8s synth" command in the CLI would generate a blank Kubernetes manifest.

Reviewer: ```suggestion
This sample shows the basic structure of a cdk8s application, which include the following components:

   * A custom `MyChart` class inherits from the [Chart](https://cdk8s.io/docs/latest/reference/cdk8s/python/#chart) 
   base class provided by the `cdk8s` library. The `Chart` base class (not to be confused with a helm chart) represents a single Kubernetes YAML manifest. 
   All resources defined within a chart will end up in the same Kubernetes manifest file.
   * An instance of the [App](https://cdk8s.io/docs/latest/reference/cdk8s/python/#app) class, which is the entry point of a cdk8s application and contains all its charts.
   * An instance of the `MyChart` class is created by passing the app instance and a string identifier, "python", as arguments. This registers the `MyChart` 
   instance (and all its resources) with the application. Note that in this example, the "python" string identifier is autogenerated based on the current directory name.
   * The `synth` method is called on the `app` instance, which produces the required Kubernetes YAML manifest files based on the defined resources. Note that in this example, 
   we haven't defined any resources within the `MyChart` constructor, so running the "cdk8s synth" command in the CLI would generate a blank Kubernetes manifest.
```

> ## Defining Kubernetes resources

Reviewer: For consistency. The two previous sections are framed as a call for action. 

```suggestion
## Define Kubernetes resources
```

Author: Updated.

> Now, we are ready to define Kubernetes resources for our application. In this sample, we illustrate the creation of a basic Kubernetes [Service](https://kubernetes.io/docs/concepts/services-networking/service/) for a sample application. We start by importing the imports directory and the **k8s** sub-directory, which contains an `__init__.py` file where all CDK8s Kubernetes classes and functions are defined.

Reviewer: ```suggestion
Now we are ready to define Kubernetes resources for our application. In this sample, we define a basic Kubernetes [Service](https://kubernetes.io/docs/concepts/services-networking/service/).
```

Author: Updated.

### Copy the code sample
1. Copy and paste the following code sample into the existing `main.py` file of your project.
```python
#!/usr/bin/env python
from cdk8s import App, Chart
from constructs import Construct
from imports import k8s

> class AwesomeApp(Chart):

Reviewer: I'd keep the code from the previous section...this got me disoriented a bit.

```suggestion
class MyChart(Chart):
```

>    def __init__(self, scope: Construct, ns: str, app_label: str):

Reviewer: I wouldn't start right away with accepting arguments to the chart. This is a somewhat more advanced usage.

```suggestion
   def __init__(self, scope: Construct, ns: str):
```

Author: I expect that developers using CDK8s are advanced developers. Biggest benefit from a programming perspective is dynamically setting labels, which we ought to highlight in some capacity.

       super().__init__(scope, ns)

       # Define a Kubernetes Service

>        k8s.KubeService(self, "my-service",

Reviewer: Defining a Service without a deployment is kind of silly. Since we don't want to overcomplicate this sample with multiple resources, how about we just create a `KubeDeployment` instead? At least that would sort of make sense.

Author: Agree, updated.

>        k8s.KubeService(self, "my-service",

Reviewer: A while back we made the `k8s` classes available inside the cdk8s+ package - but didn't change the docs to use it. Now you do:

```python
import cdk8s_plus_26 as kplus

kplus.k8s.KubeService(...)
```

This is the recommended way now actually - maybe we can adjust? This would make it so we don't have to get into the weeds of the `imports` directory and how it is created

Author: We're not referencing cdk8s+ in this particular tutorial, but let me know if there's some other pythonic logic you wanted to interject.

                       spec=k8s.ServiceSpec(
                           selector={"app": app_label},
                           ports=[k8s.ServicePort(
                               port=80, target_port=k8s.IntOrString.from_number(8080))],
                           type="LoadBalancer"
                       )
                    )

app = App()

> AwesomeApp(app, "getting-started", app_label="my-app")

Reviewer: ```suggestion
MyChart(app, "getting-started", app_label="my-app")
```

Author: We can change the class name, trying to keep it informal/friendly. But we should be introducing the idea that custom names are possible.

app.synth()
```

> A few things worth noting about this sample:

Reviewer: I think this section should just speak about the `KubeService` class:

- Where does it come from? (i.e the `k8s` directory, and what else does it contain)
- What does it map to exactly? (i.e a resource of `kind: Service`)

> - An `AwesomeApp` class is created by passing the app instance, a string identifier "getting-started", and a custom `app_label` ("my-app") as arguments. This action generates and registers the Kubernetes resources defined in the `AwesomeApp` class within the application. We use the custom `app_label` parameter to dynamically set the "app" key in the "label" selectors and "metadata" labels for our Kubernetes resources.

Reviewer: This is repeating what was already said when we first introduced the empty sample. I think we can remove it...

> - The `__init__` method of the `AwesomeApp` class creates a Kubernetes [Service](https://kubernetes.io/docs/concepts/services-networking/service/) by instantiating the `k8s.KubeService` class with the following parameters: `self` as the scope of the Service resource within the `AwesomeApp` class, a string identifier "my-service", and a `k8s.ServiceSpec` object that defines the Service specification with a selector dictionary having the key "app" and the value set to the `app_label` parameter, a list of ports containing a single `k8s.ServicePort` object with `port` set to 80 and `target_port` set to 8080, and a `type` parameter with the value "LoadBalancer". The `app_label` parameter is passed to the `AwesomeApp` class.

Reviewer: This is hard to follow. Its basically repeating exactly what the code does and I don't think adds much value. We expect people to know python, and know a bit of Kubernetes...I think we can just drop this.

Author: See above comment.

> ## Generating Kubernetes manifests

Reviewer: ```suggestion
## Generate Kubernetes manifests
```

Author: Updated.

> After you have defined the Kubernetes resources for your project, you are ready to generate the Kubernetes manifest that will define your Service resource. 

Reviewer: ```suggestion
After you have defined the Kubernetes resources for your application, you are ready to generate the Kubernetes manifest that will define your `Service` resource. 
```

Author: Updated.

### Run the synth command
1. Open a terminal and navigate to your project directory.

> 2. Run the [synth](https://cdk8s.io/docs/latest/cli/synth/) command. This command generates a Kubernetes manifest file in the `dist` folder of your project directory. The manifest file is based on the parameters specified in your application.

Reviewer: ```suggestion
2. Run the [synth](https://cdk8s.io/docs/latest/cli/synth/) command. This command generates a Kubernetes manifest file in the `dist` folder of your project directory. The manifest file contains all the resources you defined inside the `MyChart` class.
```

Author: Updated.

```console
cdk8s synth
```

### View the manifest
1. Open the `dist/getting-started.k8s.yaml` file. You should see a Kubernetes manifest similar to the following:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: getting-started-my-service-c84dc095
spec:
  ports:
    - port: 80
      targetPort: 8080
  selector:
    app: my-app
  type: LoadBalancer
```

## Wrap up

> Keep in mind that this is just a basic demonstration of CDK8s.

Reviewer: Wondering if we should add a summary sentence like "in this guide we created a cdk8s python app, defined a kuberentes service with code, and generated a YAML manifest from it"

Author: Updated. Added a conclusion to describe the Python-specific programming language conventions.

> Keep in mind that this is just a basic demonstration of CDK8s.

Reviewer: ```suggestion
Keep in mind that this is just a basic demonstration of cdk8s.
```

Author: Updated.

> - To continue your journey, develop a complete Kubernetes [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) and [Service](https://kubernetes.io/docs/concepts/services-networking/service/) using the [CDK8s Core](https://github.com/cdk8s-team/cdk8s-examples/blob/main/python/cdk8s-core/main.py) sample application.

Reviewer: ```suggestion
To continue your journey, develop a complete Kubernetes [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) and [Service](https://kubernetes.io/docs/concepts/services-networking/service/) using the [cdk8s](https://github.com/cdk8s-team/cdk8s-examples/blob/main/python/cdk8s-core/main.py) sample application.
```

Author: Updated.

> - To continue your journey, develop a complete Kubernetes [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) and [Service](https://kubernetes.io/docs/concepts/services-networking/service/) using the [CDK8s Core](https://github.com/cdk8s-team/cdk8s-examples/blob/main/python/cdk8s-core/main.py) sample application.

Reviewer: This is a bit confusing because it kind of alludes to another guide, but then just links to some code. Is this intentional or a place holder for something?

# Getting started with cdk8s for Python
In this guide, we'll walk you through the most essential concepts you need to know in order to build a Kubernetes application using cdk8s for Python.

## Prerequisites
1. To install the cdk8s CLI, you need the [Node Package Manager (npm)](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm) installed on your local machine. 
2. cdk8s for Python supports Python versions 3.7+. Check your Python version:
```bash
python3 --version
```

## Set up an environment
### Initial setup
To get setup, follow these general steps:

1. To begin, you’ll need pipenv 2018.11.26+ installed. If you’re not familiar with this setup, use these instructions to [setup pipenv](https://pipenv.pypa.io/en/latest/install/).

### Install the CLI
To initialize a cdk8s project and auto-generate Kubernetes manifests based on our code, we need the cdk8s CLI:

1. Run the following command to install the CLI using [npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm). For more installation methods, see [Install the cdk8s CLI](./../cli/installation.md).
```console
npm install -g cdk8s-cli
```

### Create a project

> Next, we’ll initialize a project to create the directory structure and install the necessary dependencies using the [init](https://cdk8s.io/docs/latest/cli/init/) command. 

Reviewer: ```suggestion
Next, we’ll initialize a project to create the directory structure and install the necessary dependencies using the [init](./../cli/init.md) command. 
```
